arangod/RestServer/DatabaseFeature.cpp

```diff
@@ -1384,3 +1393,65 @@ void DatabaseFeature::enableDeadlockDetection() {
     vocbase->_deadlockDetector.enabled(true);
   }
 }
+
+std::pair<std::string /*sharding*/, std::size_t /*replication*/>
+arangodb::getOneShardOptions(std::string const& collectionName, VPackSlice options) {
+  // Invalid options will be silently ignored. Default values will be used
+  // instead.
+  //
+  // This Function will be called twice - the second time we do not run in
+  // the risk of consulting the ClusterFeature, because defaults are provided
+  // during the first call.
+
+  TRI_ASSERT(!collectionName.empty());
+  bool systemCol = TRI_vocbase_t::IsSystemName(collectionName);
+  std::size_t replicationFactor = systemCol ? 2 : 1; // real default will be read
+                                                     // from ClusterFeature if possible
+  std::string sharding;
+
+  //  sanitize input for vocbase creation
+  //  sharding -- must be "", "flexible" or "single"
+  //  replicationFactor must be "satellite" or a natural number
+  auto shardingSlice = options.get(StaticStrings::Sharding);
+  if(! (shardingSlice.isString()  &&
+        (shardingSlice.compareString("") == 0 || shardingSlice.compareString("flexible") == 0 || shardingSlice.compareString("single") == 0)
+       )) {
+    shardingSlice = VPackSlice::noneSlice();
+  } else {
+    sharding = shardingSlice.copyString();
+  }
+
+  VPackSlice replicationSlice = options.get(StaticStrings::ReplicationFactor);
+  bool isSatellite = (replicationSlice.isString() && replicationSlice.compareString(StaticStrings::Satellite) == 0 );
+  bool isNumber = (replicationSlice.isNumber() && replicationSlice.getUInt() > 0 );
+  if(!isSatellite && !isNumber){
+    // no valid value - ignore value and use default replication factor
+    replicationSlice = VPackSlice::noneSlice();
+  } else if (isSatellite) {
+    replicationFactor = 0;
+  } else if (isNumber) {
+    replicationFactor = replicationSlice.getUInt();
+  }
+
+  // get the replication default values from cluster feature if available
+  if (replicationSlice.isNone()) {
+    ClusterFeature* cluster = dynamic_cast<ClusterFeature*>(ApplicationServer::lookupFeature("Cluster"));
```

Comment:
the extra `dynamic_cast` is not required here, because:
```
   template <typename T>
   static T* lookupFeature(std::string const& name) {
     typedef typename std::enable_if<std::is_base_of<ApplicationFeature, T>::    value, T>::type type;
     return dynamic_cast<type*>(lookupFeature(name));
   }
```